libvast/src/command.cpp

```diff
@@ -0,0 +1,258 @@
+/******************************************************************************
+ *                    _   _____   __________                                  *
+ *                   | | / / _ | / __/_  __/     Visibility                   *
+ *                   | |/ / __ |_\ \  / /          Across                     *
+ *                   |___/_/ |_/___/ /_/       Space and Time                 *
+ *                                                                            *
+ * This file is part of VAST. It is subject to the license terms in the       *
+ * LICENSE file found in the top-level directory of this distribution and at  *
+ * http://vast.io/license. No part of VAST, including this file, may be       *
+ * copied, modified, propagated, or distributed except according to the terms *
+ * contained in the LICENSE file.                                             *
+ ******************************************************************************/
+
+#include "vast/command.hpp"
+
+#include "vast/logger.hpp"
+
+namespace vast {
+
+command::command() : parent_(nullptr) {
+  // nop
+}
+
+command::command(command* parent, std::string_view name)
+  : parent_(parent),
+    name_(name) {
+  // nop
+}
+
+command::~command() {
+  // nop
+}
+
+int command::run(caf::actor_system& sys, opt_map& options, caf::message args) {
+  CAF_LOG_TRACE(CAF_ARG(options) << CAF_ARG(args));
+  // Split the arguments.
+  auto [local_args, subcmd, subcmd_args] = separate_args(args);
+  // Parse arguments for this command.
+  auto res = local_args.extract_opts(opts_);
+  if (res.opts.count("help") != 0) {
+    std::cout << res.helptext << std::endl;
+    if (!nested_.empty()) {
+      std::cout << "\nSubcommands:\n";
+      for (auto& kvp : nested_)
+        std::cout << "  " << kvp.first << "\n";
+    }
+    std::cout << std::endl;
+    return EXIT_SUCCESS;
+  }
+  // Only forward unparsed arguments to run_impl.
+  local_args = res.remainder;
+  // Populate the map with our values.
+  for (auto& f : get_opts_)
+    options.emplace(f());
+  // Check whether the options allow for further processing.
+  switch (proceed(sys, options, local_args)) {
+    default:
+        // nop
+        break;
+    case stop_successful:
+      return EXIT_SUCCESS;
+    case stop_with_error:
+      return EXIT_FAILURE;
+  }
+  // Invoke run_impl if no subcommand was defined.
+  if (subcmd.empty()) {
+    VAST_ASSERT(subcmd_args.empty());
+    return run_impl(sys, options, std::move(local_args));
+  }
+  // Dispatch to subcommand.
+  auto i = nested_.find(subcmd);
+  if (i == nested_.end()) {
+    std::cerr << "no such command: " << full_name() << " " << subcmd
+              << std::endl
+              << std::endl;
```

Comment:
How about decoupling rendering to `std::cerr` from reporting an error message? Then we can ship the error message if needed. This could be achieved by returning `caf::error`:

```cpp
auto result = cmd.run(...);
return result ? extract_return_code(result.error()) : 0;
```

Reply:
The goal of my current work is to refactor the current `run_*` functions buried in VAST's main. I think once that's done we should gather around a whiteboard to specify how we want to deal with output of commands in general (log, console, and error). I already had started something along the lines of adding `std::ostream` references to the commands, but decided against it for now to focus on the refactoring.

Reply:
Okay, if you think that refactoring this afterwards makes it easier, then let's do that first.